### PR TITLE
feat(rest_api): custom client for specific resources

### PR DESCRIPTION
**Pull request by joscha**
### Description

This adds the ability to pass a custom `ClientConfig` to an `EndpointResource`.
The passed `ClientConfig` object will be merged with the global one from  the containing `@dlt.resource`.

#### Example

```py
{
    "client": {
        "base_url": "https://api.mything.com",
        # Default auth is using Bearer authentication
        "auth": {
            "type": "bearer",
            "token": dlt.secrets["your_api_token"],
        }
    },
    "resources": [
        "resource-using-bearer-auth", # <-- for example this resource
        {
            "name": "my-resource-with-special-auth",
            "client": {
                # However this resource is using HTTP Basic auth instead.
                "auth": HttpBasicAuth("user", dlt.secrets["your_basic_auth_password"]) 
            },
            # ...
        }
    ]
    # ...
}
```

In the corresponding [Slack thread](https://dlthub-community.slack.com/archives/C04DQA7JJN6/p1732109635628149) a question came up whether a custom authenticator could be used. The answer is yes, it would be possible to write custom authenticator which introspects the resource requested (base on the URL for example), however the implementation of such an authenticator would be a lot more involved, less obvious than just merging a configuration object and it would also not have an answer for differing `base_url`s , `headers` or other client-specific parameters for example.

### Related Issues

- Fixes #...
- Closes #...
- Resolves #...

### Additional Context

See originating [Slack thread](https://dlthub-community.slack.com/archives/C04DQA7JJN6/p1732109635628149).

The reason for this feature is that some API endpoints, even though available under the same base URL and returning the same entities, may have different ways of accessing them. For example the [Affinity V2 API](https://developer.affinity.co/) uses an OpenAPI-based access model with [Bearer authentication](https://developer.affinity.co/#section/Getting-Started/Authentication). This part of the API is available under `https://api.affinity.co/v2/`.
The [first version of the API](https://api-docs.affinity.co/) is available under `https://api.affinity.co/` but uses [HTTP Basic Auth](https://api-docs.affinity.co/#authentication).
Entities from both APIs are using the same entity IDs and are related, thus must be specified in the same `@dlt.resource` in order to define those relations.
This is currently (`1.4.0`) not possible, due to the differing (incompatible) authentication methods.

### CODE REVIEW

File: dlt/sources/rest_api/__init__.py
Comment (joscha) [inline dlt/sources/rest_api/__init__.py:266]: configs get merged here, with the keys defined in the endpoint taking preference over the global one.

File: dlt/sources/rest_api/typing.py
Comment (joscha) [inline dlt/sources/rest_api/typing.py:212]: Typing overrides are currently not yet supported (see https://github.com/python/mypy/issues/7435) thus the ignore here.
For the optional dict in each endpoint we need to ensure that all keys are optional.

File: tests/sources/rest_api/configurations/source_configs.py
Comment (joscha) [inline tests/sources/rest_api/configurations/source_configs.py:406]: example of a valid config with an `auth` override, such as what is needed for solving the issue described in the pull request.

File: dlt/sources/rest_api/__init__.py
Comment (francescomucio) [inline dlt/sources/rest_api/__init__.py:414]: if you do `c.get("resources", [])`, you can skip the `if resources:`

File: dlt/sources/rest_api/__init__.py
Comment (francescomucio) [inline dlt/sources/rest_api/__init__.py:417]: I would go for the opposite check (is instance of `EndpointResource`?), is there a reason for testing like this

Comment (joscha) [reply dlt/sources/rest_api/__init__.py:417]: Yes, it's a `TypedDict` and the check is not (yet) supported on it :-/

Comment (francescomucio) [reply dlt/sources/rest_api/__init__.py:417]: sure, you could test only for `dict`, but that should be enough. if's it's a `dict` (even if an incorrect one) the code will try to mask the client field. Then the real validation will happen.

It is not super precise (a comment to mention that the actual test should be for the proper type, but python...), but a bit simpler to read.

Happy to hear other opinions @burnash

File: dlt/sources/rest_api/__init__.py
Comment (joscha) [inline dlt/sources/rest_api/__init__.py:418]:
```suggestion
    resources = c.get("resources", [])
    for resource in resources:
        if isinstance(resource, str) or isinstance(resource, DltResource):
            continue
```

File: dlt/sources/rest_api/__init__.py
Comment (joscha) [reply dlt/sources/rest_api/__init__.py]: changed in 1bf1fe766ab071f6a5df70f9434e0ae2b8c984ba

**Comment by joscha** [comment]
@willi-mueller @burnash would you mind taking a look and give me feedback on this, please?

**Comment by burnash** [comment]
Hi @joscha, thanks you for the contribution. I think this is a great idea. As you may have noticed, `rest_api` already creates a `RESTClient` instance for each of the resources under the hood.

One thing that I would like to discuss is that the PR extends to the resource configuration with the `client` key which, in turn, creates an additional way to configure the `paginator`. This is already configured via `endpoint`:

```python
{
    "client": {
        "base_url": "https://api.mything.com",
        "auth": {
            # ...
        },
        "paginator": "json_link"
    },
    "resources": [
        "resource-using-bearer-auth",
        {
            "name": "my-resource-with-special-auth",
            "client": {
                "auth": HttpBasicAuth("user", dlt.secrets["your_basic_auth_password"]),
                "paginator": "json_link"  # <-- now it could be configured here as well
            },
            "endpoint": {
                "paginator": "json_link"
            }
        }
    ]
}
```

I think it would be better to keep the configuration in one place. To achieve this, I see two options:
1. Move the `paginator` configuration to the resources's `client` from the `endpoint` configuration.
2. Do not add a `client` configuration to resources, but extend the `endpoint` configuration to accept `auth`.

Although I think the first option is more consistent with the global `client` configuration, I would prefer the second option because:
1. It is shorter for cases where the user needs to configure pagination (which happens often).
2. It is backward-compatible.

Your example would then look like this:
```python
{
    "client": {
        "base_url": "https://api.mything.com",
        "auth": {
            "type": "bearer",
            "token": dlt.secrets["your_api_token"],
        }
    },
    "resources": [
        "resource-using-bearer-auth",
        {
            "name": "my-resource-with-special-auth",
            "endpoint": {
                "path": "my-resource-with-special-auth",
                "auth": HttpBasicAuth("user", dlt.secrets["your_basic_auth_password"]),
                "paginator": "json_link",
            }
        }
    ]
}
```

What do you think?

**Comment by joscha** [comment]
Hi @burnash,

I initially had only `auth` as it was enough for what I wanted to do initially but when I opened the PR I noticed that this would not actually allow a completely reconfigured rest client. Some APIs when migrating between versions may choose to use different header parameters and/or base_urls so I thought overwriting/merging the whole client object would be best, as this would first of all work for all eventualities but also in concept be more simple than just add one separate object and then later another one, etc. whilst having to stay backwards compatible.
What do you think about removing the paginator as a possible property for overrides of the client parameters? Would that be an option? I do generally think your outlined 1. is cleaner long-term. If we were to accept both we could get the benefit of the short notation. Given the endpoint is more deeply nested I think the behavior would be that the endpoint paginator would override resource paginator would override global client paginator.
It is quite some baggage to carry though, for the convenience of omitting one extra `client {}`.

**Comment by francescomucio** [comment]
For my experience, changing the paginator is something more common than changing the authentication method for different endpoints of the same API.

I am in favour of the first option, but I can see how it can cause some confusion. In my opinion we should make clear how things are handled, possible options are:

1. the client in the resource cannot contain a client.paginator, but only an endpoint.paginator
2. the client in the resource can contain a client.paginator, but the endpoint.paginator wins.

So it will be like

`client.paginator` < `default_resources.client.paginator` < `default_resource.endpoint.paginator` < `resource.client.paginator` < `resource.endpoint.paginator`

What do you think?

**Comment by burnash** [comment]
Thank you @joscha and @francescomucio for your thoughtful input and discussion on this PR. After considering all the points raised, I wanted to share my perspective.

I'm not fully convinced that adding a client configuration directly to the resource is the optimal approach. In practice, setting the paginator explicitly per endpoint is much more common than changing the authentication method within the same API. Introducing a client config at the resource level may lead to unnecessary ambiguity.

Regarding the headers, I believe these should be specified within the endpoint configuration. In fact, there's already a PR adding headers at the endpoint level https://github.com/dlt-hub/dlt/pull/2084 

As for the `base_url`, I'm not sure it's necessary within the resource configuration since `endpoint.path` can accept an absolute URL (I'd need update the documentation to reflect this). Putting headers and auth into the endpoint configuration would make it consistent with how the underlying `RESTClient` and `RESTClient.paginate` operate.

Here's my proposal for moving forward. 
1. Extend the endpoint configuration to accept `auth` which overrides the client's auth configuration for the endpoint.
2. Add headers configuration to the endpoint via https://github.com/dlt-hub/dlt/pull/2084 

Here's how the configuration would look:

```python
{
    "client": {
        "base_url": "https://api.mything.com",
        "auth": {
            "type": "bearer",
            "token": dlt.secrets["your_api_token"],
        }
    },
    "resources": [
        "resource-using-bearer-auth",
        {
            "name": "my-resource-with-special-auth",
            "endpoint": {
                "path": "my-resource-with-special-auth",
                "auth": HttpBasicAuth("user", dlt.secrets["your_basic_auth_password"]),
                "paginator": "json_link",
                "headers": {"Custom-Header": "value"}
            }
        }
    ]
}
```

@joscha, would you be open to updating the PR to reflect this approach?

**Comment by joscha** [comment]
I can most definitely update the PR that way if that's what you wish. I think possibly spreading the individual properties of `Client` into the endpoint config is shortsighted as it will lead to an explosion of highly specialized properties spread into that object of whenever another is missing to support the next person's use case.
But I haven't DLT long enough to say whether that is likely or not, so given that, unless I hear back from you otherwise, I will update the PR later to use auth and headers.
I will also update the docs with the mention of the absolute URL as it is a related issue, I was not aware this was accepted before you wrote it.
Thank you!

**Comment by joscha** [comment]
Purely in regards to mechanics, do you:
* Want me to wait until the open headers PR is decided and merged
* Open the PR anyway and we deal with conflicts
* Want me to base my PR off the headers PR

?

**Comment by burnash** [comment]
Thank you @joscha for your understanding and flexibility! Let's proceed with adding just the auth configuration to the endpoint config for now. Since #2084 is already handling the headers implementation. I will help with the managing the conflicts.

**Comment by joscha** [comment]
changes in @burnash. Decided to do the mention of being able to use an absolute URL in a separate pull request.

**Comment by joscha** [comment]
I just saw that `verified-sources` seems to have a copy (?) of the rest client - see
https://github.com/dlt-hub/verified-sources/blob/3d2993a685e8c1d713a2283ac4025ca3903ed392/sources/rest_api/__init__.py#L249

1. is that intentional? (my assumption is yes, although none of the other verified sources seem to use it) 
2. is there a process that syncs the two? If yes, where can I find it? if no, why not?
3. if 2.) is "no", should I make the same change in this repository as well?

File: docs/website/docs/dlt-ecosystem/verified-sources/rest_api/basic.md
Comment (burnash) [inline docs/website/docs/dlt-ecosystem/verified-sources/rest_api/basic.md:311]: I think it should be `client` not, resource, e.g.:
```suggestion
- `auth`: An optional `AuthConfig` instance. If passed, is used over the one defined in the [client](#client) definition. Example:
```

File: dlt/sources/rest_api/__init__.py
Comment (burnash) [inline dlt/sources/rest_api/__init__.py:415]: I think this can be simplified to:
```suggestion
        if isinstance(resource, (str, DltResource)):
```

File: dlt/sources/rest_api/__init__.py
Comment (burnash) [inline dlt/sources/rest_api/__init__.py:418]: And this to:

```suggestion
        if endpoint := resource.get("endpoint"):
```

File: dlt/sources/rest_api/__init__.py
Comment (burnash) [inline dlt/sources/rest_api/__init__.py:412]: Unrelated to this PR: there's not much point in this assignment, but I'll fix that in a follow up PR

**Review by burnash — CHANGES_REQUESTED** [review]
Thanks you @joscha 
Look good, there are a couple of small improvements needed, please the comments.

**Comment by burnash** [comment]
Very good question.

> I just saw that `verified-sources` seems to have a copy (?) of the rest client - see https://github.com/dlt-hub/verified-sources/blob/3d2993a685e8c1d713a2283ac4025ca3903ed392/sources/rest_api/__init__.py#L249
> 
> 1. is that intentional? (my assumption is yes, although none of the other verified sources seem to use it)

Yes, it's intentional, we moved  rest_api, sql_database and filesystem sources to dlt core in [1.0.0](https://github.com/dlt-hub/dlt/releases/tag/1.0.0). Versions prior to 1.0.0 still reference these sources as being in the verified-sources repository when you execute `dlt init`.

> 2. is there a process that syncs the two? If yes, where can I find it? if no, why not?

There isn’t a formal sync mechanism in place. If someone is working with projects that require or benefit from the latest updates or additional features available in the newer versions of these resources, I recommend upgrading to dlt version 1.0.0 or later.

> 3. if 2.) is "no", should I make the same change in this repository as well?

As this is a feature PR, for now, there's no need to port these changes to verified-sources repo. For bugfixes it's different.

File: tests/sources/rest_api/integration/test_response_actions.py
Comment (joscha) [inline tests/sources/rest_api/integration/test_response_actions.py:324]:
```suggestion
        assert (
            response.request.headers["Authorization"]
            == f"Basic {base64.b64encode(b'U:P').decode('ascii')}"
        )
```

File: dlt/sources/rest_api/__init__.py
Comment (joscha) [reply dlt/sources/rest_api/__init__.py]: removed in e2f6c8cf

File: tests/sources/rest_api/configurations/source_configs.py
Comment (burnash) [inline tests/sources/rest_api/configurations/source_configs.py:404]: Looks like this line is causing the tests to fail as `{repository}` path parameter is missing.
```suggestion
                    "path": "user/repos",
```

Comment (joscha) [reply tests/sources/rest_api/configurations/source_configs.py]: ah, thank you!

**Comment by joscha** [comment]
🎉 all good to go now

**Review by burnash — APPROVED** [review]
> 🎉 all good to go now

Yes, looks great, thank you!

**Comment by joscha** [comment]
Awesome, thank you. Are there instructions on how to use the current devel branch at a given sha? The only mention of nightly builds I could find is here: https://github.com/dlt-hub/dlt/issues/997

**Comment by burnash** [comment]
You can use [git+ prefix in pip](https://pip.pypa.io/en/stable/topics/vcs-support/), for example:

Install a devel branch:
```
pip install "git+https://github.com/dlt-hub/dlt.git@devel"
```

Install a devel branch with duckdb extra
```
pip install "git+https://github.com/dlt-hub/dlt.git@devel#egg=dlt[duckdb]"
```

Replace `devel` with an sha to get a specific commit.

**Pull request merged**
